// @flatfile/api/dist/api/resources/events/types/EventTopic.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventTopic = void 0;
exports.EventTopic = {
    AgentCreated: "agent:created",
    AgentUpdated: "agent:updated",
    AgentDeleted: "agent:deleted",
    SpaceCreated: "space:created",
    SpaceUpdated: "space:updated",
    SpaceDeleted: "space:deleted",
    DocumentCreated: "document:created",
    DocumentUpdated: "document:updated",
    DocumentDeleted: "document:deleted",
    WorkbookCreated: "workbook:created",
    WorkbookUpdated: "workbook:updated",
    WorkbookDeleted: "workbook:deleted",
    SheetCreated: "sheet:created",
    SheetUpdated: "sheet:updated",
    SheetDeleted: "sheet:deleted",
    SnapshotCreated: "snapshot:created",
    RecordsCreated: "records:created",
    RecordsUpdated: "records:updated",
    RecordsDeleted: "records:deleted",
    FileCreated: "file:created",
    FileUpdated: "file:updated",
    FileDeleted: "file:deleted",
    JobCreated: "job:created",
    JobUpdated: "job:updated",
    JobDeleted: "job:deleted",
    JobCompleted: "job:completed",
    JobReady: "job:ready",
    JobScheduled: "job:scheduled",
    JobOutcomeAcknowledged: "job:outcome-acknowledged",
    JobFailed: "job:failed",
    CommitCreated: "commit:created",
    CommitUpdated: "commit:updated",
    CommitCompleted: "commit:completed",
    LayerCreated: "layer:created",
};
