// @flatfile/api/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlatfileClient = void 0;
const Client_1 = require("./api/resources/spaces/client/Client");
const Client_2 = require("./api/resources/agents/client/Client");
const Client_3 = require("./api/resources/auth/client/Client");
const Client_4 = require("./api/resources/cells/client/Client");
const Client_5 = require("./api/resources/documents/client/Client");
const Client_6 = require("./api/resources/environments/client/Client");
const Client_7 = require("./api/resources/events/client/Client");
const Client_8 = require("./api/resources/files/client/Client");
const Client_9 = require("./api/resources/guests/client/Client");
const Client_10 = require("./api/resources/jobs/client/Client");
const Client_11 = require("./api/resources/records/client/Client");
const Client_12 = require("./api/resources/sheets/client/Client");
const Client_13 = require("./api/resources/users/client/Client");
const Client_14 = require("./api/resources/versions/client/Client");
const Client_15 = require("./api/resources/workbooks/client/Client");
class FlatfileClient {
    constructor(options) {
        this.options = options;
    }
    get spaces() {
        var _a;
        return ((_a = this._spaces) !== null && _a !== void 0 ? _a : (this._spaces = new Client_1.Spaces(this.options)));
    }
    get agents() {
        var _a;
        return ((_a = this._agents) !== null && _a !== void 0 ? _a : (this._agents = new Client_2.Agents(this.options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new Client_3.Auth(this.options)));
    }
    get cells() {
        var _a;
        return ((_a = this._cells) !== null && _a !== void 0 ? _a : (this._cells = new Client_4.Cells(this.options)));
    }
    get documents() {
        var _a;
        return ((_a = this._documents) !== null && _a !== void 0 ? _a : (this._documents = new Client_5.Documents(this.options)));
    }
    get environments() {
        var _a;
        return ((_a = this._environments) !== null && _a !== void 0 ? _a : (this._environments = new Client_6.Environments(this.options)));
    }
    get events() {
        var _a;
        return ((_a = this._events) !== null && _a !== void 0 ? _a : (this._events = new Client_7.Events(this.options)));
    }
    get files() {
        var _a;
        return ((_a = this._files) !== null && _a !== void 0 ? _a : (this._files = new Client_8.Files(this.options)));
    }
    get guests() {
        var _a;
        return ((_a = this._guests) !== null && _a !== void 0 ? _a : (this._guests = new Client_9.Guests(this.options)));
    }
    get jobs() {
        var _a;
        return ((_a = this._jobs) !== null && _a !== void 0 ? _a : (this._jobs = new Client_10.Jobs(this.options)));
    }
    get records() {
        var _a;
        return ((_a = this._records) !== null && _a !== void 0 ? _a : (this._records = new Client_11.Records(this.options)));
    }
    get sheets() {
        var _a;
        return ((_a = this._sheets) !== null && _a !== void 0 ? _a : (this._sheets = new Client_12.Sheets(this.options)));
    }
    get users() {
        var _a;
        return ((_a = this._users) !== null && _a !== void 0 ? _a : (this._users = new Client_13.Users(this.options)));
    }
    get versions() {
        var _a;
        return ((_a = this._versions) !== null && _a !== void 0 ? _a : (this._versions = new Client_14.Versions(this.options)));
    }
    get workbooks() {
        var _a;
        return ((_a = this._workbooks) !== null && _a !== void 0 ? _a : (this._workbooks = new Client_15.Workbooks(this.options)));
    }
}
exports.FlatfileClient = FlatfileClient;
