// @flatfile/api/dist/serialization/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.workbooks = exports.versions = exports.users = exports.spaces = exports.snapshots = exports.sheets = exports.secrets = exports.records = exports.property = exports.mapping = exports.jobs = exports.guests = exports.files = exports.events = exports.environments = exports.documents = exports.commons = exports.commits = exports.cells = exports.auth = exports.agents = void 0;
exports.agents = __importStar(require("./agents"));
__exportStar(require("./agents/types"), exports);
exports.auth = __importStar(require("./auth"));
__exportStar(require("./auth/types"), exports);
exports.cells = __importStar(require("./cells"));
__exportStar(require("./cells/types"), exports);
exports.commits = __importStar(require("./commits"));
__exportStar(require("./commits/types"), exports);
exports.commons = __importStar(require("./commons"));
__exportStar(require("./commons/types"), exports);
exports.documents = __importStar(require("./documents"));
__exportStar(require("./documents/types"), exports);
exports.environments = __importStar(require("./environments"));
__exportStar(require("./environments/types"), exports);
exports.events = __importStar(require("./events"));
__exportStar(require("./events/types"), exports);
exports.files = __importStar(require("./files"));
__exportStar(require("./files/types"), exports);
exports.guests = __importStar(require("./guests"));
__exportStar(require("./guests/types"), exports);
exports.jobs = __importStar(require("./jobs"));
__exportStar(require("./jobs/types"), exports);
exports.mapping = __importStar(require("./mapping"));
__exportStar(require("./mapping/types"), exports);
exports.property = __importStar(require("./property"));
__exportStar(require("./property/types"), exports);
exports.records = __importStar(require("./records"));
__exportStar(require("./records/types"), exports);
exports.secrets = __importStar(require("./secrets"));
__exportStar(require("./secrets/types"), exports);
exports.sheets = __importStar(require("./sheets"));
__exportStar(require("./sheets/types"), exports);
exports.snapshots = __importStar(require("./snapshots"));
__exportStar(require("./snapshots/types"), exports);
exports.spaces = __importStar(require("./spaces"));
__exportStar(require("./spaces/types"), exports);
exports.users = __importStar(require("./users"));
__exportStar(require("./users/types"), exports);
exports.versions = __importStar(require("./versions"));
__exportStar(require("./versions/types"), exports);
exports.workbooks = __importStar(require("./workbooks"));
__exportStar(require("./workbooks/types"), exports);
__exportStar(require("./files/client/requests"), exports);
__exportStar(require("./records/client/requests"), exports);
__exportStar(require("./snapshots/client/requests"), exports);
__exportStar(require("./users/client/requests"), exports);
__exportStar(require("./versions/client/requests"), exports);
