// @flatfile/api/serialization/resources/billing/types/StripeProduct.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StripeProduct = void 0;
const core = __importStar(require("../../../../core"));
exports.StripeProduct = core.serialization.object({
    id: core.serialization.string(),
    object: core.serialization.string(),
    active: core.serialization.boolean(),
    attributes: core.serialization.list(core.serialization.string()),
    created: core.serialization.number(),
    defaultPrice: core.serialization.property("default_price", core.serialization.string()),
    description: core.serialization.string(),
    images: core.serialization.list(core.serialization.string()),
    livemode: core.serialization.boolean(),
    metadata: core.serialization.unknown(),
    name: core.serialization.string(),
    packageDimensions: core.serialization.property("package_dimensions", core.serialization.string()),
    shippable: core.serialization.boolean(),
    statementDescriptor: core.serialization.property("statement_descriptor", core.serialization.string()),
    taxCode: core.serialization.property("tax_code", core.serialization.string()),
    type: core.serialization.string(),
    unitLabel: core.serialization.property("unit_label", core.serialization.string()),
    updated: core.serialization.number(),
    url: core.serialization.string(),
});
