// @flatfile/plugin-record-hook/dist/index.mjs
// src/record.translater.ts
import {
  FlatfileRecord,
  FlatfileRecords
} from "@flatfile/hooks";
var RecordTranslater = class {
  constructor(records) {
    this.records = records;
    this.toFlatFileRecords = () => {
      if (this.records instanceof FlatfileRecords) {
        return this.records;
      } else {
        const XRecords = this.records;
        const FFRecords = new FlatfileRecords(
          XRecords.map((record) => {
            let rawData = {};
            for (let [k, v] of Object.entries(record.values)) {
              if (!!v.links?.length && v.value) {
                const links = v.links.map((link) => {
                  let linkedRawData = {};
                  for (let [lk, lv] of Object.entries(link.values)) {
                    linkedRawData[lk] = lv.value;
                  }
                  return linkedRawData;
                });
                rawData[k] = {
                  value: v.value,
                  links
                };
              } else {
                rawData[k] = v.value;
              }
            }
            const metadata = record.metadata;
            const rawRecord = {
              rowId: record.id,
              rawData,
              metadata
            };
            return rawRecord;
          })
        );
        return FFRecords;
      }
    };
    this.toXRecords = () => {
      if (this.records[0] instanceof FlatfileRecord) {
        const FFRecords = this.records;
        return FFRecords.map((record) => {
          const recordWithInfo = record.toJSON();
          let values = {};
          for (let [k, v] of Object.entries(recordWithInfo.row.rawData)) {
            const messages = recordWithInfo.info.filter((info) => info.field === k).map((info) => ({ message: info.message, type: info.level }));
            values[k] = {
              value: v !== null && typeof v === "object" ? v.value : v,
              messages,
              valid: true
            };
          }
          const metadata = recordWithInfo.row.metadata;
          return {
            id: String(record.rowId),
            values,
            metadata
          };
        });
      } else {
        return this.records;
      }
    };
    this.records = records;
  }
};

// src/RecordHook.ts
import api from "@flatfile/api";
var RecordHook = async (event, handler) => {
  const { sheetId } = event.context;
  try {
    const records = await event.cache.init(
      "records",
      async () => (await event.data).records
    );
    if (!records)
      return;
    const batch = await prepareXRecords(records);
    for (const x of batch.records) {
      await handler(x, event);
    }
    const recordsUpdates = new RecordTranslater(
      batch.records
    ).toXRecords();
    await event.cache.set("records", async () => recordsUpdates);
    event.afterAll(async () => {
      const records2 = event.cache.get("records");
      try {
        return await api.records.update(sheetId, records2);
      } catch (e) {
        console.log(`Error putting records: ${e}`);
      }
    });
  } catch (e) {
    console.log(`Error getting records: ${e}`);
  }
  return handler;
};
var prepareXRecords = async (records) => {
  const clearedMessages = records.map(
    (record) => {
      Object.keys(record.values).forEach((k) => {
        record.values[k].messages = [];
      });
      return record;
    }
  );
  const fromX = new RecordTranslater(clearedMessages);
  return fromX.toFlatFileRecords();
};

// src/index.ts
var recordHook = (sheetSlug, callback) => {
  return (client) => {
    client.on(
      "commit:created",
      {
        // todo: fix this filter
        context: {
          // @ts-ignore
          sheetSlug
        }
      },
      (event) => {
        return RecordHook(event, callback);
      }
    );
  };
};
export {
  RecordHook,
  recordHook
};
